### PR TITLE
add lint id enum (#426)

### PR DESCRIPTION
**Pull request by fabricio7p**
Began refactoring `src/checks.rs`

- add child module check_variants
- add Lint variants
- refactor test setup to support the variants

Note: this is just a checkpoint to see if i'm on the right track. Code does not compile yet!

### CODE REVIEW

**Comment by mgrachev** [comment]
@fabricio7p 👋 Thank you for your contribution 👍

Let us know when you've finished so we can review your PR 😉

**Comment by fabricio7p** [comment]
Checkpoint.

- add CheckVariants struct that hold the variants vec. (Vec<Lint>)
- refactor `comment.rs` to support output the new CheckVariants struct
- begin refactoring each `/src/fixes` (only first 3 done)

**Comment by fabricio7p** [comment]
Refactor done.
Please review @mgrachev 

- Refactored every occurence of string comparisons to support the new LintKind enum.
- Refactored all tests to support enum variants
- Moved Lint and LintKind to new `lints.rs` file
- Add testing to Lint and LintKind

**Comment by mgrachev** [comment]
I'm sorry, I'm quite busy now, so I'll try to review your PR by the end of the week 🙂

File: src/lib.rs [hunk 1]
```diff
@@ -9,9 +9,11 @@ pub mod cli;
 mod common;
 mod fixes;
 mod fs_utils;
+mod lints;
```
Comment (mgrachev) [inline src/lib.rs:12]: Let it be `lint_kind`.

File: src/common/comment.rs [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 const PREFIX: &str = "dotenv-linter";
 const ON: &str = "on";
 const OFF: &str = "off";
 
 #[derive(Debug, PartialEq)]
-pub struct Comment<'a> {
+pub struct Comment {
     disable: bool,
-    pub checks: Vec<&'a str>,
+    pub checks: Lint,
```
Comment (mgrachev) [inline src/common/comment.rs:10]: Can you use just `Vec<LintKind>`?

**Review by mgrachev — COMMENTED** [review]
@fabricio7p Thank you for your contribution 🚀

I'm sorry for the late reply.
I have left some comments.
Please take a look at them 👀

File: src/checks/duplicated_key.rs [hunk 3]
```diff
@@ -1,9 +1,10 @@
 use crate::checks::Check;
 use crate::common::*;
+use crate::lints::LintKind;
 use std::collections::HashSet;
 
 pub(crate) struct DuplicatedKeyChecker<'a> {
-    name: &'a str,
+    name: LintKind,
```
[the remark this answers is not in the document]
Comment (fabricio7p) [reply src/checks/duplicated_key.rs]: Done

File: src/common/comment.rs [hunk 4]
```diff
@@ -85,15 +90,15 @@ mod tests {
         assert_eq!(
             Some(Comment {
                 disable: false,
-                checks: vec![],
+                checks: Lint::new(),
```
[the remark this answers is not in the document]
Comment (fabricio7p) [reply src/common/comment.rs]: Done

File: src/common/comment.rs [hunk 5]
```diff
@@ -123,15 +128,15 @@ mod tests {
         assert_eq!(
             Some(Comment {
                 disable: true,
-                checks: vec!["UnorderedKey"],
+                checks: Lint::from(vec!["UnorderedKey"]),
```
[the remark this answers is not in the document]
Comment (fabricio7p) [reply src/common/comment.rs]: Done

File: src/lints.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
+use std::{fmt, str::FromStr};
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Lint {
```
[the remark this answers is not in the document]
Comment (fabricio7p) [reply src/lints.rs]: Done

File: src/lib.rs (hunk 1 above)
Comment (fabricio7p) [reply src/lib.rs]: Done

File: src/common/comment.rs (hunk 2 above)
Comment (fabricio7p) [reply src/common/comment.rs]: Done

File: src/fixes.rs [hunk 7]
```diff
@@ -133,7 +137,7 @@ mod tests {
         ];
         let mut warnings = vec![Warning::new(
             lines[1].clone(),
-            "Unfixable",
+            LintKind::Unfixable,
```
[the remark this answers is not in the document]
Comment (fabricio7p) [reply src/fixes.rs]: Done

File: src/lints.rs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    IncorrectDelimiter,
+    KeyWithoutValue,
+    LeadingCharacter,
+    LowercaseKey,
+    QuoteCharacter,
+    SpaceCharacter,
+    SubstitutionKey,
+    TrailingWhitespace,
+    UnorderedKey,
+    Unfixable,
```
[the remark this answers is not in the document]
Comment (fabricio7p) [reply src/lints.rs]: Done

**Comment by fabricio7p** [comment]
Alright, everything refactored.
Rebased from master as well.

File: src/common/comment.rs [hunk 9]
```diff
@@ -52,6 +63,13 @@ impl Comment<'_> {
 mod tests {
     use super::*;
 
+    fn as_lintkind_vec(vec: Vec<&str>) -> Vec<LintKind> {
```
Comment (mgrachev) [inline src/common/comment.rs:66]: I don't think that we need a special function in the tests module to convert vectors.
Just use `vec![LintKind::UnorderedKey]`.

File: src/common/line_entry.rs [hunk 10]
```diff
@@ -301,6 +301,16 @@ mod tests {
     }
 
     mod get_control_comment {
+        use crate::lint_kind::LintKind;
+        use std::str::FromStr;
+
+        fn as_lintkind_vec(vec: Vec<&str>) -> Vec<LintKind> {
```
Comment (mgrachev) [inline src/common/line_entry.rs:307]: I don't think that we need a special function in the tests module to convert vectors.
Just use `vec![LintKind::UnorderedKey]`.

File: src/lib.rs [hunk 11]
```diff
@@ -62,6 +71,11 @@ pub fn fix(args: &clap::ArgMatches, current_dir: &Path) -> Result<(), Box<dyn Er
     if let Some(skip) = args.values_of("skip") {
         skip_checks = skip.collect();
     }
+    let skip_checks = skip_checks
+        .into_iter()
+        .map(LintKind::from_str)
+        .collect::<Result<Vec<LintKind>, _>>()
+        .unwrap();
```
Comment (mgrachev) [inline src/lib.rs:78]: https://github.com/dotenv-linter/dotenv-linter/pull/427/files#r660372735

File: src/lib.rs [hunk 12]
```diff
@@ -28,6 +31,12 @@ pub fn check(args: &clap::ArgMatches, current_dir: &Path) -> Result<usize, Box<d
         skip_checks = skip.collect();
     }
 
+    let skip_checks = skip_checks
+        .into_iter()
+        .map(LintKind::from_str)
+        .collect::<Result<Vec<LintKind>, _>>()
+        .unwrap();
```
Comment (mgrachev) [inline src/lib.rs:38]: https://github.com/dotenv-linter/dotenv-linter/pull/427/files#r660372735

**Comment by fabricio7p** [comment]
All good.
Tests passing :+1:

**Review by mgrachev — APPROVED** [review]

**Comment by mgrachev** [comment]
@fabricio7p Thank you!

Please squash your commits to one and fix warnings
https://github.com/dotenv-linter/dotenv-linter/pull/427/checks?check_run_id=2950521632

**Review by DDtKey — APPROVED** [review]
Thank you for very useful PR 🙂

**Comment by mgrachev** [comment]
@fabricio7p Thank you for your contribution 👍

**Pull request merged**